gossip/src/crds_value.rs

```diff
@@ -490,84 +494,166 @@ impl Sanitize for NodeInstance {
     }
 }
 
-#[derive(Serialize, Deserialize, Clone, Default, PartialEq, Eq, AbiExample, Debug)]
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, AbiExample, AbiEnumVisitor)]
+enum SlotsOffsets {
+    RunLengthEncoding(RunLengthEncoding),
+    RawOffsets(RawOffsets),
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+struct U16(#[serde(with = "serde_varint")] u16);
+
+// The vector always starts with 1. Encode number of 1's and 0's consecutively.
+// For example, 110000111 is [2, 4, 3].
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+struct RunLengthEncoding(Vec<U16>);
+
+impl RunLengthEncoding {
+    fn new(bits: &BitVec<u8>) -> Self {
+        let encoded = (0..bits.len())
+            .map(|i| bits.get(i))
+            .dedup_with_count()
+            .map_while(|(count, _)| u16::try_from(count).ok())
+            .scan(0, |current_bytes, count| {
+                *current_bytes += ((u16::BITS - count.leading_zeros() + 6) / 7).max(1) as usize;
+                (*current_bytes <= RestartLastVotedForkSlots::MAX_BYTES).then_some(U16(count))
+            })
+            .collect();
+        Self(encoded)
+    }
+
+    fn num_encoded_slots(&self) -> usize {
+        self.0.iter().map(|x| usize::from(x.0)).sum::<usize>()
+    }
+
+    fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut slots: Vec<Slot> = self
+            .0
+            .iter()
+            .zip([1, 0].iter().cycle())
+            .flat_map(|(bit_count, bit)| std::iter::repeat(bit).take(usize::from(bit_count.0)))
+            .enumerate()
+            .filter(|(_, bit)| **bit == 1)
+            .map_while(|(offset, _)| {
+                let offset = Slot::try_from(offset).ok()?;
+                last_slot.checked_sub(offset)
+            })
+            .take(MAX_SLOTS_PER_ENTRY)
```

Comment:
RunLengthEncoding can potentially pack tons of zeroes in it, but we only send back 1's to the caller, and it's bound by local root, so I think you are right we can remove this.

Reply:
well, we can pack tones of ones as well, and `last_root` can be equal to `u64::MAX`, and then the sender can send a lot of slots due to RLE encoding.
so now that I am thinking, we actually need to limit the length of output vector here because you cannot trust the value.
The limit does not need to be `MAX_SLOTS_PER_ENTRY` because those are not related anymore but we do in fact need a `.take(max_slots)` here. sorry for the back and forth on this.

`RawOffsets::to_slots` does not need a limit because that is already limited by the packet size.

Reply:
Thanks for catching this. I think we can go back to 81k slots in the original design doc then. However, if it's 81k slots then u16 no longer fits, I've changed it into u32.